### PR TITLE
James 1746 Demonstration of junit-contracts

### PR DESCRIPTION
**Pull request by chibenwa**
I added the mappers' test to what we support with junit-contracts ...

Found one missing license.

Definitly, this change highlighted an inconsistent behaviour between memory and Cassandra as Memory didn't extended AbstractAttachmentMapper test and test were thus not runned. With junit-contracts, you can not ignore a part of the contract. Thus it is a better tool.

Here I demonstrate : 
 - Compatibility with (basic) rules (in the common part of the tests)
 - How to replace @Text(expected=...) where you can not rely on assertThrownBy (because version 1 of assertj)
 - Genericity, and several contracts for a single class

### CODE REVIEW

File: server/data/data-jmap-cassandra/src/test/java/org/apache/james/jmap/cassandra/access/CassandraAccessTokenRepositoryTest.java [hunk 1]
```diff
@@ -20,22 +20,34 @@
...
+import org.xenei.junit.contract.ContractImpl;
+import org.xenei.junit.contract.ContractSuite;
+import org.xenei.junit.contract.IProducer;
 
-public class CassandraAccessTokenRepositoryTest extends AbstractAccessTokenRepositoryTest {
+@RunWith(ContractSuite.class)
+@ContractImpl(CassandraAccessTokenRepository.class)
+public class CassandraAccessTokenRepositoryTest {
 
     private CassandraCluster cassandra;
```
Comment (mbaechler) [inline server/data/data-jmap-cassandra/src/test/java/org/apache/james/jmap/cassandra/access/CassandraAccessTokenRepositoryTest.java]: you should not keep that private field, don't you ?

File: mailbox/cassandra/src/test/java/org/apache/james/mailbox/cassandra/mail/CassandraMapperProviderProducer.java [hunk 2]
```diff
@@ -19,12 +19,41 @@
...
+import org.xenei.junit.contract.IProducer;
 
-public class CassandraAttachmentMapperTest extends AbstractAttachmentMapperTest {
+import com.google.common.base.Throwables;
 
-    public CassandraAttachmentMapperTest() {
-        super(new CassandraMapperProvider());
+@RunWith(ContractSuite.class)
+@ContractImpl(CassandraMapperProvider.class)
+public class CassandraMapperProviderProducer {
```
Comment (mbaechler) [inline mailbox/cassandra/src/test/java/org/apache/james/mailbox/cassandra/mail/CassandraMapperProviderProducer.java]: The class name doesn't look very logical to me.

Comment (chibenwa) [inline mailbox/cassandra/src/test/java/org/apache/james/mailbox/cassandra/mail/CassandraMapperProviderProducer.java]: Ok

File: server/data/data-jmap-cassandra/src/test/java/org/apache/james/jmap/cassandra/access/CassandraAccessTokenRepositoryTest.java [hunk 3]
```diff
@@ -20,22 +20,35 @@
...
+import org.xenei.junit.contract.ContractSuite;
+import org.xenei.junit.contract.IProducer;
 
-public class CassandraAccessTokenRepositoryTest extends AbstractAccessTokenRepositoryTest {
+@RunWith(ContractSuite.class)
+@ContractImpl(CassandraAccessTokenRepository.class)
+public class CassandraAccessTokenRepositoryTest {
 
-    private CassandraCluster cassandra;
+    private IProducer<CassandraAccessTokenRepository> producer = new IProducer<CassandraAccessTokenRepository>() {
```
Comment (mbaechler) [inline server/data/data-jmap-cassandra/src/test/java/org/apache/james/jmap/cassandra/access/CassandraAccessTokenRepositoryTest.java]: you don't need the exact type on the right

File: server/data/data-jmap/src/test/java/org/apache/james/jmap/memory/access/MemoryAccessTokenRepositoryTest.java [hunk 4]
```diff
@@ -19,12 +19,32 @@
...
+import org.xenei.junit.contract.IProducer;
 
-public class MemoryAccessTokenRepositoryTest extends AbstractAccessTokenRepositoryTest {
+@RunWith(ContractSuite.class)
+@ContractImpl(MemoryAccessTokenRepository.class)
+public class MemoryAccessTokenRepositoryTest {
 
-    protected MemoryAccessTokenRepository createAccessTokenRepository() {
-        return new MemoryAccessTokenRepository(TTL_IN_MS);
+    private IProducer<MemoryAccessTokenRepository> producer = new IProducer<MemoryAccessTokenRepository>() {
```
Comment (mbaechler) [inline server/data/data-jmap/src/test/java/org/apache/james/jmap/memory/access/MemoryAccessTokenRepositoryTest.java]: same thing about generic type

File: mailbox/pom.xml [hunk 5]
```diff
@@ -415,16 +415,9 @@
...
-                <version>${jmock.version}</version>
-                <scope>test</scope>
-            </dependency>
-            <dependency>
-                <groupId>org.jmock</groupId>
-                <artifactId>jmock-junit4</artifactId>
-                <version>${jmock.version}</version>
-                <scope>test</scope>
+                <groupId>org.xenei</groupId>
+                <artifactId>junit-contracts</artifactId>
```
Comment (mbaechler) [inline mailbox/pom.xml]: this change is probably unrelated

File: mailbox/api/src/test/java/org/apache/james/mailbox/MailboxManagerStressTest.java [hunk 6]
```diff
@@ -119,16 +137,11 @@ public void run() {
...
-            } else {
-                last = uid;
-            }
-
-        }
+        // check if there is no duplicates
+        // For mailboxes without locks, even if the UID is monotic, as re-scheduling can happen between UID generation and event delivery,
+        // we can not check the order on the event listener
+        // No UID duplicates prevents message loss
+        assertEquals(APPEND_OPERATIONS, ImmutableSet.copyOf(uList).size());
```
Comment (mbaechler) [inline mailbox/api/src/test/java/org/apache/james/mailbox/MailboxManagerStressTest.java]: could you split this part ?

File: mailbox/pom.xml (hunk 5 above)
Comment (chibenwa) [inline mailbox/pom.xml]: This change IS related. Overwise, junit-contracts breaks. Conflicting dependency on JMock.

File: mailbox/api/src/test/java/org/apache/james/mailbox/MailboxManagerStressTest.java (hunk 6 above)
Comment (chibenwa) [inline mailbox/api/src/test/java/org/apache/james/mailbox/MailboxManagerStressTest.java]: In a separate commit ? Sure I can.

File: mailbox/pom.xml (hunk 5 above)
Comment (mbaechler) [inline mailbox/pom.xml]: I don't really understand why, but ok.

File: server/data/data-jmap-cassandra/src/test/java/org/apache/james/jmap/cassandra/access/CassandraAccessTokenRepositoryTest.java (hunk 3 above)
Comment (chibenwa) [inline server/data/data-jmap-cassandra/src/test/java/org/apache/james/jmap/cassandra/access/CassandraAccessTokenRepositoryTest.java]: If I put AccessTokenRepository either on the right or the left of the "=" I 've got an Incompatible type compilation error in IntelliJ.

So I'll keep it like this....

File: server/data/data-jmap-cassandra/src/test/java/org/apache/james/jmap/cassandra/access/CassandraAccessTokenRepositoryTest.java [hunk 7]
```diff
@@ -31,7 +32,7 @@
 @ContractImpl(CassandraAccessTokenRepository.class)
 public class CassandraAccessTokenRepositoryTest {
 
-    private IProducer<CassandraAccessTokenRepository> producer = new IProducer<CassandraAccessTokenRepository>() {
+    private IProducer<AccessTokenRepository> producer = new IProducer<AccessTokenRepository>() {
```
Comment (mbaechler) [inline server/data/data-jmap-cassandra/src/test/java/org/apache/james/jmap/cassandra/access/CassandraAccessTokenRepositoryTest.java]: sorry but it's more about using the diamond operator : 
`private IProducer<AccessTokenRepository> producer = new IProducer<>() {`

Comment (chibenwa) [inline server/data/data-jmap-cassandra/src/test/java/org/apache/james/jmap/cassandra/access/CassandraAccessTokenRepositoryTest.java]:
```
Unchecked assignment: 'anonymous org.xenei.junit.contract.IProducer' to 'org.xenei.junit.contract.IProducer<org.apache.james.jmap.ServerWrapper>' less... (Ctrl+F1) 
Signals places where an unchecked warning is issued by the compiler, for example:

  void f(HashMap map) {
    map.put("key", "value");
  }
  
Hint: Pass -Xlint:unchecked to javac to get more details....
```

**Comment by mbaechler** [comment]
:+1:

**Comment by aduprat** [comment]
:+1:

**Pull request closed without merging**